spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala

```diff
@@ -17,80 +17,352 @@
+      || !isTableDVFree(snapshot)
+      || numFiles == 0
+      || !statsMinMaxNullColumns.columns.contains("minValues")
+      || !statsMinMaxNullColumns.columns.contains("maxValues")
+      || !statsMinMaxNullColumns.columns.contains("nullCount")) {
+      return (Some(numRecords), Map.empty)
+    }
+
+    // dataColumns can contain columns without stats if dataSkippingNumIndexedCols
+    // has been increased
+    val columnsWithStats = files.select(
+      col("stats.minValues.*"),
+      col("stats.maxValues.*"),
+      col("stats.nullCount.*"))
+      .columns.groupBy(identity).mapValues(_.size)
+      .filter(x => x._2 == 3) // 3: minValues, maxValues, nullCount
+      .map(x => x._1).toSet
+
+    // Creates a tuple with physical name to avoid recalculating it multiple times
+    val dataColumnsWithStats = dataColumns.map(x => (x, DeltaColumnMapping.getPhysicalName(x)))
+      .filter(x => columnsWithStats.contains(x._2))
+
+    val columnsToQuery = dataColumnsWithStats.flatMap { columnAndPhysicalName =>
+      val dataType = columnAndPhysicalName._1.dataType
+      val physicalName = columnAndPhysicalName._2
+
+      Seq(col(s"stats.minValues.`$physicalName`").cast(dataType).as(s"min.$physicalName"),
+        col(s"stats.maxValues.`$physicalName`").cast(dataType).as(s"max.$physicalName"),
+        col(s"stats.nullCount.`$physicalName`").as(s"nullCount.$physicalName"))
+    } ++ Seq(col(s"stats.numRecords").as(s"numRecords"))
+
+    val minMaxExpr = dataColumnsWithStats.flatMap { columnAndPhysicalName =>
+      val physicalName = columnAndPhysicalName._2
+
+      // To validate if the column has stats we do two validation:
+      // 1-) COUNT(nullCount.columnName) should be equals to numFiles,
+      // since nullCount is always non-null.
+      // 2-) The number of files with non-null min/max:
+      // a. count(min.columnName)|count(max.columnName) +
+      // the number of files where all rows are NULL:
+      // b. count of (ISNULL(min.columnName) and nullCount.columnName == numRecords)
+      // should be equals to numFiles
+      Seq(
+        s"""case when $numFiles = count(`nullCount.$physicalName`)
+            | AND $numFiles = (count(`min.$physicalName`) + sum(case when
+            |  ISNULL(`min.$physicalName`) and `nullCount.$physicalName` = numRecords
+            |   then 1 else 0 end))
+            | AND $numFiles = (count(`max.$physicalName`) + sum(case when
+            |  ISNULL(`max.$physicalName`) AND `nullCount.$physicalName` = numRecords
+            |   then 1 else 0 end))
+            | then TRUE else FALSE end as `complete_$physicalName`""".stripMargin,
+        s"min(`min.$physicalName`) as `min_$physicalName`",
+        s"max(`max.$physicalName`) as `max_$physicalName`")
+    }
+
+    val statsResults = files.select(columnsToQuery: _*).selectExpr(minMaxExpr: _*).head
+
+    (Some(numRecords), dataColumnsWithStats
+      .filter(x => statsResults.getAs[Boolean](s"complete_${x._2}"))
```

Comment:
If stats are not complete for one of the requested columns, should we just return the empty stats? Can this rule use stats available for partial set of columns referenced in the agg expressions?

Reply:
Currently it can't use partial set of columns. I'm not doing the extra check here to avoid running the query twice for success case